collections/permutations.ts

```diff
@@ -20,45 +20,43 @@
  * ])
  * ```
  */
-export function permutations<T>(array: readonly T[]): T[][] {
+export function permutations<T>(inputArray: readonly T[]): T[][] {
   const ret: T[][] = [];
+  const k = inputArray.length;
 
-  if (array.length === 0) {
+  if (k === 0) {
     return ret;
   }
 
-  // Heap Algorithm
-  function heapPermutations(k: number, array: T[]) {
-    const c = new Array<number>(k).fill(0);
+  // Heap's Algorithm
+  const array = [...inputArray];
+  const c = new Array<number>(k).fill(0);
 
-    ret.push([...array]);
+  ret.push([...array]);
 
-    let i = 1;
+  let i = 1;
 
-    while (i < k) {
-      if (c[i] < i) {
-        if (i % 2 === 0) {
-          const swapBuffer = array[0];
-          array[0] = array[i];
-          array[i] = swapBuffer;
-        } else {
-          const swapBuffer = array[c[i]];
-          array[c[i]] = array[i];
-          array[i] = swapBuffer;
-        }
-
-        ret.push([...array]);
-
-        c[i] += 1;
-        i = 1;
+  while (i < k) {
+    if (c[i] < i) {
+      if (i % 2 === 0) {
+        const swapBuffer = array[0];
+        array[0] = array[i];
+        array[i] = swapBuffer;
       } else {
-        c[i] = 0;
-        i += 1;
+        const swapBuffer = array[c[i]];
+        array[c[i]] = array[i];
+        array[i] = swapBuffer;
```

Comment:
Missed this on the my implementation, but this could be

```suggestion
        [array[c[i]], array[i]] = [array[i], array[c[i]]];
```

instead

Reply:
I think we could keep the current implementation since we are into performance and memory optimisation.

Reply:
I thought the destructuring syntax gives v8 the chance to optimize (as the expression on the right can not be referenced outside of the statement), while it would need a very specific optimization for the buffer version to be faster.

I could not find details in any of the two with a quick search though...

Reply:
I have updated with array destructuring.

My quick search suggests that it will likely be good enough nowadays, and my quick *likely flawed* benchmark with 10 elements suggest that the performance overhead is around 3% after a thousand iterations.

I am fine either way since `permutations` is already a slow O(n!) operation :)

Some fun readings: 
https://v8.dev/blog/v8-release-68#array-destructuring-improvements 
https://docs.google.com/document/d/1hWb-lQW4NSG9yRpyyiAA_9Ktytd5lypLnVLhPX9vamE/edit#heading=h.j2s8yv21fevz